docs/indexing.rst

```diff
@@ -0,0 +1,139 @@
+    in the batch.
+
+Indexing from the end
+~~~~~~~~~~~~~~~~~~~~~
+
+Negative indices can be used to index the tensor starting from the end. The index of -1 denotes
+the last element::
+
+    channels = sizes[-1]   # channels go last
+    widths = sizes[-2]     # widths are the innermost dimension after channels
+
+Indexing with run-time values
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Indexing by constant is often insufficient. With DALI, you can use a result of other computations
+to access tensor elements. In the example below, we use a run-time defined index to access
+an element at a random position within a tensor::
+
+    raw_files = fn.readers.file(...)
+    length = fn.shapes(raw_files)[0]
+
+    # calculate a random index from 0 to file_length-1
+    random_01 = fn.random.uniform(range=(0, 1))  # random numbers in range [0..1)
+    index = fn.floor(random_01 * length)  # calculate indices from [0..length)
+    index = fn.cast(index, dtype=dali.types.INT64)  # cast the index to integer - required for indexing
+
+    # extract a random byte
+    random_byte = raw_files[index]
+
+Here, a byte at random index will be extracted from each sample. ``index`` is a data node which
+represents a batch of scalar values, one per sample. Each of these values is used as the index
+for the respective sample in ``raw_files``.
+
+.. note::
+    The index must be a result of a CPU operator.
+
+
+Slicing
+~~~~~~~
+
+To extract multiple values (or slices), the Python list slicing systax can be used::
+
+    header = raw_files[:16]  # extract 16-byte headers from files in the batch
+
+If the start of the slice is omitted, the slice starts at 0. If the end is omitted, the slice
+ends at the end of given axis. Negative indices can be used for both start and end of the slice.
+Either end can be a constant, a run-time value (a DataNode) or can be skipped.
+
+Multidimensional selection
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The slection can contain multiple axes, separated by comma. If an index is a scalar, the
+corresponding dimension is removed from the output::
+
+    images = fn.decoders.image(jpegs, device="mixed")  # RGB images in HWC layout
+    red =   images[:,:,0]
+    green = images[:,:,1]
+    blue =  images[:,:,2]
+
+The ``red``, ``green``, ``blue`` are 2D tensors.
+
+Slicing keeps the sliced dimensions even if the length of the slice is 1::
+
+    green_with_channel = images[:,:,1:2]  # the last dimension is kept
+
+When indexing and slicing multidimensional data, the trailing dimenions can be omitted. This is
+equivalent to passing a full-range slice to all trailing dimensions::
+
+    wide = letterboxed[20:-20,:,:]   # slice height, keep width and channels
+    wide = letterboxed[20:-20,:]     # this line is equivalent to the previous one
+    wide = letterboxed[20:-20]       # ...and so is this one
+
+.. note::
+    See also :func:`fn.crop` and :func:`fn.slice` for operations tailored for image processing.
```

Comment:
Maybe it would be useful to add very small code example on how you can replace slice with indexing somewhere in this doc?